src/platform/ESP32/ESP32EndpointQueueFilter.h

```diff
@@ -0,0 +1,128 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <inet/EndpointQueueFilter.h>
+#include <lwip/ip4_addr.h>
+
+namespace chip {
+namespace Inet {
+
+class ESP32EndpointQueueFilter : public EndpointQueueFilter
+{
+public:
+    CHIP_ERROR SetMdnsHostName(const chip::CharSpan & hostName)
+    {
+        ReturnErrorCodeIf(hostName.size() != sizeof(mHostNameBuffer), CHIP_ERROR_INVALID_ARGUMENT);
+        ReturnErrorCodeIf(!IsValidMdnsHostName(hostName), CHIP_ERROR_INVALID_ARGUMENT);
+        memcpy(mHostNameBuffer, hostName.data(), hostName.size());
+        return CHIP_NO_ERROR;
+    }
+
+    FilterOutcome FilterBeforeEnqueue(const void * endpoint, const IPPacketInfo & pktInfo,
+                                      const chip::System::PacketBufferHandle & pktPayload) override
+    {
+        if (!IsMdnsBroadcastPacket(pktInfo))
+        {
+            return FilterOutcome::kAllowPacket;
+        }
+        // Drop the mDNS packets which don't contain 'matter' or '<device-hostname>'.
+        const uint8_t matterBytes[] = { 'm', 'a', 't', 't', 'e', 'r' };
+        if (PayloadContains(pktPayload, ByteSpan(matterBytes)) || PayloadContainsHostNameCaseInsensitive(pktPayload))
+        {
+            return FilterOutcome::kAllowPacket;
+        }
+        return FilterOutcome::kDropPacket;
+    }
+
+    FilterOutcome FilterAfterDequeue(const void * endpoint, const IPPacketInfo & pktInfo,
+                                     const chip::System::PacketBufferHandle & pktPayload) override
+    {
+        return FilterOutcome::kAllowPacket;
+    }
+
+private:
+    // TODO: Add unit tests for these static functions
+    static bool IsMdnsBroadcastPacket(const IPPacketInfo & pktInfo)
+    {
+#if INET_CONFIG_ENABLE_IPV4
+        if (pktInfo.DestAddress.IsIPv4() && pktInfo.DestPort == 5353)
+        {
+            ip4_addr_t mdnsBroadcastAddr4;
+            ip4addr_aton("224.0.0.251", &mdnsBroadcastAddr4);
+            return pktInfo.DestAddress.ToIPv4().addr == mdnsBroadcastAddr4.addr;
+        }
+#endif
+        if (pktInfo.DestAddress.IsIPv6() && pktInfo.DestPort == 5353)
+        {
+            ip6_addr_t mdnsBroadcastAddr6;
+            ip6addr_aton("ff02::fb", &mdnsBroadcastAddr6);
```

Comment:
Suggest making a constant for ff02::fb without `aton`, since aton is costly, if you keep the address. Still think the address check itselt is not needed.

Reply:
Use the condition `pktInfo.DestAddress == mdns::Minimal::BroadcastIpAddresses::Get(chip::Inet::IPAddressType::kIPv6)`

Reply:
Done